amqp/src/main/scala/akka/stream/alpakka/amqp/impl/AmqpFlow.scala

```diff
@@ -0,0 +1,185 @@
+    extends GraphStageWithMaterializedValue[FlowShape[WriteMessage[T], WriteResult[T]], Future[Done]] { stage =>
+
+  private val in: Inlet[WriteMessage[T]] = Inlet(Logging.simpleName(this) + ".in")
+  private val out: Outlet[WriteResult[T]] = Outlet(Logging.simpleName(this) + ".out")
+
+  override val shape: FlowShape[WriteMessage[T], WriteResult[T]] = FlowShape.of(in, out)
+
+  override protected def initialAttributes: Attributes =
+    Attributes.name(Logging.simpleName(this)) and ActorAttributes.IODispatcher
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+    (new AbstractAmqpFlowStage[T](writeSettings, promise, shape) {
+      override def publish(message: WriteMessage[T]): Unit = {
+        log.debug("Publishing message {}.", message)
+
+        channel.basicPublish(
+          writeSettings.exchange.getOrElse(""),
+          message.routingKey.orElse(writeSettings.routingKey).getOrElse(""),
+          message.mandatory,
+          message.immediate,
+          message.properties.orNull,
+          message.bytes.toArray
+        )
+        push(out, WriteResult.confirmed(message.passThrough))
+      }
+    }, promise.future)
+  }
+}
+
+/**
+ * Internal API.
+ *
+ * AMQP flow that waits for confirmation after every published manner in a blocking way.
+ * Messages are emitted as failed in case when `channel.waitForConfirms` either returns `false`, or throws
+ * [[java.util.concurrent.TimeoutException]]. In case of any other exception during publication or waiting for
+ * confirmation the stage is failed.
+ */
+@InternalApi private[amqp] final class AmqpBlockingFlow[T](
+    writeSettings: AmqpWriteSettings,
+    confirmationTimeout: FiniteDuration
+) extends GraphStageWithMaterializedValue[FlowShape[WriteMessage[T], WriteResult[T]], Future[Done]] { stage =>
+
+  private val in: Inlet[WriteMessage[T]] = Inlet(Logging.simpleName(this) + ".in")
+  private val out: Outlet[WriteResult[T]] = Outlet(Logging.simpleName(this) + ".out")
+
+  override val shape: FlowShape[WriteMessage[T], WriteResult[T]] = FlowShape.of(in, out)
+
+  override protected def initialAttributes: Attributes =
+    Attributes.name(Logging.simpleName(this)) and ActorAttributes.IODispatcher
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val promise = Promise[Done]()
+    (new AbstractAmqpFlowStage[T](writeSettings, promise, shape) {
+
+      override def whenConnected(): Unit =
+        channel.confirmSelect()
+
+      override def publish(message: WriteMessage[T]): Unit = {
+        val publicationResult: Try[Boolean] = for {
+          _ <- publishWithBlockingConfirm(message)
+          result <- waitForConfirmation()
+        } yield result
+
+        log.debug("Publication result: {}", publicationResult)
+
+        publicationResult
```

Comment:
looks too complex and long, could it be simplified?

I think everything could be done in one `Future#transform` or `Future#transformWith`...

Reply:
I moved timeout handling code inside `waitForConfirmation` as it seems to better fit there, and used fold instead of transform - code looks way better now. If you have further suggestions i'm open 😉